url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/gui/quantum/Visualizer.java

```diff
@@ -0,0 +1,547 @@
+		for (int i = 0; i < qubit.length; i++)
+			if (qubit[i].getType() == Qubit.type.input)
+				n_input++;
+			else
+				break;
+		
+		if (n_input > 0) {
+			final int coeff_to_gen = 1 << n_input;
+			
+			char first = 'a';
+			sb.append(first);
+			
+			for (int i = 1; i < coeff_to_gen; i++) {
+				first++;
+				sb.append(", "+first);
+			}
+			
+			desc.inputs = sb.toString();
+			sb.setLength(0);
+		}
+		
+		// set entanglement
+		final Entangler[] entanglers = getAllItemsOfType(new Entangler[0]);
+		boolean firstitem = true;
+		for (final Entangler e : entanglers) {
+			final String text = "("+Tools.zBOB(e.i1.getId())+", "+Tools.zBOB(e.i2.getId())+")";
+			if (firstitem) {
+				sb.append(text);
+				firstitem = false;
+			} else
+				sb.append("; "+text);
+		}
+		desc.entanglement = sb.toString();
+		sb.setLength(0);
+		
+		// set measurements
+		firstitem = true;
+		for (final Qubit q : qubit)
+			if (q.perform_measurement) {
+				final String text = "("+Tools.zBOB(q.getId())+", "+getDependance(q, corrtype.X, "s")+", "+getDependance(q, corrtype.Z, "t")+", "+q.measurement_angle+")";
+				if (firstitem) {
+					sb.append(text);
+					firstitem = false;
+				} else
+					sb.append("; "+text);
+			}
+		desc.measurements = sb.toString();
+		sb.setLength(0);
+		
+		// corrections
+		firstitem = true;
+		for (final Qubit q : qubit)
+			if (!q.perform_measurement) {
+				
+				final String xdep = getDependance(q, corrtype.X, "s");
+				final String zdep = getDependance(q, corrtype.Z, "s");
+
+				final String textx = xdep.equals("0") ? null : "("+Tools.zBOB(q.getId())+", x, "+xdep+")";
+				final String textz = zdep.equals("0") ? null : "("+Tools.zBOB(q.getId())+", z, "+zdep+")";
+
+
+				if (textx != null) {
+					if (firstitem) {
+						sb.append(textx);
+						firstitem = false;
+					} else
+						sb.append("; "+textx);
+				}
+
+				if (textz != null) {
+					if (firstitem) {
+						sb.append(textz);
+						firstitem = false;
+					} else
+						sb.append("; "+textz);
+				}
+			}
+		desc.corrections = sb.toString();
+		sb.setLength(0);
+		
+		return desc;
+	}
+		
+	private int getOccurances(final String key, final HashMap<String, Integer> db) {
+		final Integer count = db.get(key);
+		return count == null ? 0 : count;
+	}
+	
+	public void saveToFile(final File f) throws Exception {
+		// for storing how many times each class have been seen
+		final HashMap<String, Integer> db = new HashMap<String, Integer>();
+		
+		for (final Item it : items) {
+			final String its = it.saveToString();
+			if (its != null) {
+				final String className = it.getClass().getName();
+				final Integer occurances = getOccurances(className, db);
+				
+				p.setProperty(className+"&"+occurances, its);
+				
+				db.put(className, occurances == null ? 1 : (occurances + 1));
+			}
+		}
+		
+		// save occurances
+		for (final String key : db.keySet())
+			p.setProperty(key, String.valueOf(db.get(key)));
+		
+		FileOutputStream out;
+		out = new FileOutputStream(f);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at line 504. The resource is closed at line 505. There are other execution paths that do not contain closure statements, for example, when Properties.store throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).